tests/merch.test.ts

```diff
@@ -438,3 +440,109 @@ describe('merch item options', () => {
       .toStrictEqual(item.options[0].getPublicMerchItemOption(true));
   });
 });
+
+describe('merch order pickup events', () => {
+  test('GET /order/pickup/future returns pickup events that haven\'t ended yet', async () => {
+    const conn = await DatabaseConnection.get();
+    const admin = UserFactory.fake({ accessType: UserAccessType.ADMIN });
+    const pastPickupEvent = MerchFactory.fakeOrderPickupEvent({
+      start: moment().subtract(2, 'hour').toDate(),
+      end: moment().subtract(1, 'hour').toDate(),
+    });
+    const ongoingPickupEvent = MerchFactory.fakeOrderPickupEvent({
+      start: moment().subtract(30, 'minutes').toDate(),
+      end: moment().add(30, 'minutes').toDate(),
+    });
+    const futurePickupEvent = MerchFactory.fakeOrderPickupEvent({
+      start: moment().add(1, 'hour').toDate(),
+      end: moment().add(2, 'hour').toDate(),
+    });
+
+    await new PortalState()
+      .createUsers(admin)
+      .createOrderPickupEvents(pastPickupEvent, ongoingPickupEvent, futurePickupEvent)
+      .write();
+
+    const getFuturePickupEventsResponse = await ControllerFactory.merchStore(conn).getFuturePickupEvents(admin);
+    expect(getFuturePickupEventsResponse.pickupEvents).toEqual(
+      expect.arrayContaining([
+        ongoingPickupEvent.getPublicOrderPickupEvent(true),
+        futurePickupEvent.getPublicOrderPickupEvent(true),
+      ]),
+    );
+  });
+
+  test('POST /order/pickup persists pickup event on proper input', async () => {
+    const conn = await DatabaseConnection.get();
+    const admin = UserFactory.fake({ accessType: UserAccessType.ADMIN });
+    const pickupEvent = MerchFactory.fakeOrderPickupEvent();
+
+    await new PortalState()
+      .createUsers(admin)
+      .write();
+
+    const merchStoreController = ControllerFactory.merchStore(conn);
+    await merchStoreController.createPickupEvent({ pickupEvent }, admin);
+
+    const [persistedPickupEvent] = await conn.manager.find(OrderPickupEventModel, { relations: ['orders'] });
```

Comment:
We only have GET for future pickup events, so using database lookups I think is the only way to actually pull all persisted events.

Reply:
Also, I think we should standardize how these tests are operating on our system. My understanding's that these are 'API unit tests', meaning they're testing each API route as an individual unit, and that we'd eventually have integration tests for testing E2E operations like signing up -> verifying -> attending an event -> viewing leaderboard or logging in -> visit merch store -> select few items -> verify cart -> place order -> select pickup event.

If each test is testing an individual unit, then I don't think we should be using any other API routes than the one we are testing, so that a bug in 1 controller wouldn't affect multiple tests, which makes it easier to debug tests. To me it makes sense to use database lookups since those we have full control over what we're querying (which matters in this case), and so that we aren't relying on the implementation of existing lookups which might be buggy.

Reply:
> If each test is testing an individual unit, then I don't think we should be using any other API routes than the one we are testing, so that a bug in 1 controller wouldn't affect multiple tests, which makes it easier to debug tests. To me it makes sense to use database lookups since those we have full control over what we're querying (which matters in this case), and so that we aren't relying on the implementation of existing lookups which might be buggy.

Agreed but GETs tend to be so simple I think it's fine and even preferable to use those instead of queries.

> Also, I think we should standardize how these tests are operating on our system. My understanding's that these are 'API unit tests', meaning they're testing each API route as an individual unit, and that we'd eventually have integration tests for testing E2E operations like signing up -> verifying -> attending an event -> viewing leaderboard or logging in -> visit merch store -> select few items -> verify cart -> place order -> select pickup event.

Can't we do this by calling the controllers like we do now, except in larger tests that more closely resemble actual user flows? I don't know how much work it'd be to setup some other system that launches the backend app and then makes real HTTP requests, but it doesn't seem worth it based on test coverage. What would that test that our current setup doesn't? Everything outside of the controller methods, so (1) ensuring HTTP requests work, (2) the API routes are valid (e.g. `GET /event/:uuid` doesn't "cover" `GET /event/future`, where a request to `GET /event/future` is evaluated as `GET /event/:uuid` where `uuid=future`), (3) the routes correctly map to controller methods, (4) request validation.

I think those are all important, but we can assume routing-controllers works as intended for (1) and (3), and I think we can write tests for (2) and (4). (2) maybe by doing something similar to the migrations test, which parses files into ASTs and checks the timestamps in the class names, except the routes test would parse controllers into annotations and verify that no route "covers" any subsequent route using regexes, and (4) by tests that manually validate.

Related to (4), something I've been thinking about wrt API specs and the API client we talked about a while back is if we can have all our controller methods defined as interfaces that the controllers then implement, and one way of doing that would be to accept raw requests in the controller methods and then validate in the method. That'd clean up our method signatures a bit too, since we could do something like
```typescript
// interface
function editMerchItem(uuid, MerchItemEdit);

// controller
function editMerchItem(@Param uuid: any, merchItemEdit: any) {
  CommonValidators.validateUuid(uuid);
  MerchStoreValidators.validateMerchItemEdit(merchItemEdit);
}
```
Don't know if that even works though, and it's not really that relevant to this discussion, except that then our existing tests would also test validation.